File: Java/Behavioral/Memento/TextEditor.java
/** Originator - creates and restores mementos */
public class TextEditor {
    private String content;

    public void write(String text) {
        content = text;
        System.out.println("Current content: " + content);
    }

    public Memento save() {
        System.out.println("Saving state...");
        return new Memento(content);
    }

    public void restore(Memento memento) {
        content = memento.getState();
        System.out.println("Restored content: " + content);
    }

    public String getContent() {
        return content;
    }
}

File: Javascript/Behavioral/Memento/text-editor.js
/**
 * Memento Pattern - Text Editor with Undo
 */

class EditorMemento {
  constructor(content) {
    this.content = content;
    this.timestamp = new Date();
  }

  getContent() {
    return this.content;
  }
}

class TextEditor {
  constructor() {
    this.content = '';
  }

  type(text) {
    this.content += text;
  }

  getContent() {
    return this.content;
  }

  save() {
    return new EditorMemento(this.content);
  }

  restore(memento) {
    this.content = memento.getContent();
  }
}

class History {
  constructor() {
    this.mementos = [];
  }

  push(memento) {
    this.mementos.push(memento);
  }

  pop() {
    return this.mementos.pop();
  }
}

module.exports = { TextEditor, History };
